Feat: Set operations (union / intersect / except) support

Angelelz opened the pull request:
This PR will close #207

Angelelz committed:
(Mysql) Initial implementation of Set Operations support

Added a new file set-operators.ts with a SetOperatorQueryBuilder class and a SetOperator class

The MySqlSelectBuilder now extends the SetOperatorQueryBuilder to inherit the set operator methods

Added the export for the set operation functions

Angelelz committed:
Initial implementation of Set Operations support for postgres

This implementation will likely change with the same pattern used in the Mysql implementation

Angelelz committed:
Progress on type for only union in Mysql. More work needed

Angelelz committed:
Added generic types to mysql Set Operation to generate type errors if the user attempts to select different types on both sides of the set operator

Angelelz committed:
Added support to pass multiple arguments to the standalone set operator functions. The types might need some more work

Angelelz committed:
Added function support to the set operators methods in mysql

Angelelz committed:
(Mysql) fixed set operator function accepting only one parameter

Angelelz committed:
(mysql) added type tests for the set operators

Angelelz committed:
[MySql] fixed type issue with the rest parameters passed to the set operator functions

 - Added a new type 'SetOperatorRestSelect' to properly handle the rest parameter

 - Added a new generic parameter TRest to properly handle the rest parameter

 - Added additional type tests for the rest parameters

 - deleted debigging strings on ValidateShape type

Angelelz committed:
Added test for mysql implementation of set operators

 - fixed wring import in pg that prevented test from running

 - added tests

Angelelz committed:
[Pg] Completed implementation of set operations

 - Added abstract class PgSetOperatorBuilder that extends the TypedQueryBuilder

 - Added type helpers for correct type inference including when passing multiple parameters to the function form

 - PgSelect now extends from PgSetOperatorBuilder instead of TypedQueryBuilder to inherit all its methods

Angelelz committed:
[Pg] Added type tests for the set operators

Angelelz committed:
[Pg] Added integration tests for the set operators

Angelelz committed:
[SQLite] Completed implementation of set operators

 - added SQLiteSetOperatorBuilder and SQLiteSetOperator classes

 - SQLiteSelectBuilder now extends from SQLiteSetOperatorBuilder to inherit its methods

 - Exported funciton versions of union, unionAll, intersect and except

 - Had to add the last generic parameter as any to the in a type for libsql driver

Angelelz committed:
[SQLite] Added type tests

Angelelz committed:
[SQLite] Added integration tests for all cases

Angelelz commented:
In its most basic form, upon merging this PR, a drizzle user will be able to create queries with union / intersect / except {all} statements with the following syntax:
```ts
await db.select().from(table)
  .union(
    db.select().from(table)
  );
  ```
  
Beyond that, given the following queries:
```ts
const query1 = db
  .select({ id: users.id, name: users.name })
  .from(users)
  .where(eq(users.id, 1))
  .leftJoin(posts, eq(users.id, posts.authorId));

const query2 = db
  .select({ id: users.id, name: sql<string>`users.name` })
  .from(users)
  .where(eq(users.id, 2));

const query3 = db
  .select({ id: users.id, name: users.name })
  .from(users)
  .where(eq(users.id, 3));

const query4 = db
  .select({ id: users.id, name: sql<string>`users.name` })
  .from(users)
  .where(eq(users.id, 4))
  .leftJoin(posts, eq(users.id, posts.authorId));

const query5 = db
  .select({ id: users.id, name: users.name })
  .from(users)
  .where(eq(users.id, 5));

const query6 = db
  .select({ id: users.id, name: users.name })
  .from(users)
  .where(eq(users.id, 2));
```

All of the following are valid drizzle syntax:
```ts
const result1 = await union(
  query1,
  query2,
  intersect(query3, query4, query5),
  query6,
).orderBy(desc(users.id))

const result2 = await union(
  union(
    union(query1, query2),
    intersect(intersect(query3, query4), query5),
  ),
  query6,
)

const result3 = await query1
  .union(query2)
  .union(query3.intersect(query4).intersect(query5))
  .union(query6)

const result4 = await query1
  .union(query2)
  .union(({ intersect }) => intersect(query3, query4, query5))
  .union(query6)
```

And will all result in the same query (MySql and Postgres):
Keep in mind that the queries are mutable, so if those 4 queries are run one after the other, the results would be unexpected.
```sql
"(((select \"users\".\"id\", \"users\".\"name\" from \"users\" left join \"posts\" on \"users\".\"id\" = \"posts\".\"author_id\"
where \"users\".\"id\" = $1) union (select \"id\", users.name from \"users\" where \"users\".\"id\" = $2))
union (((select \"id\", \"name\" from \"users\" where \"users\".\"id\" = $3) intersect (select \"users\".\"id\", users.name
from \"users\" left join \"posts\" on \"users\".\"id\" = \"posts\".\"author_id\" where \"users\".\"id\" = $4))
intersect (select \"id\", \"name\" from \"users\" where \"users\".\"id\" = $5)))
union (select \"id\", \"name\" from \"users\" where \"users\".\"id\" = $6)",
params: [ 1, 2, 3, 4, 5, 2 ]
```

But in SQLite, parenthesis are not allowed, or order by / limit on the left of the set operator clause, so the resulting query will be:
```sql
"select \"users\".\"id\", \"users\".\"name\" from \"users\" left join \"posts\" on \"users\".\"id\" = \"posts\".\"author_id\"
where \"users\".\"id\" = ? union select \"id\", users.name from \"users\" where \"users\".\"id\" = ?
union select \"id\", \"name\" from \"users\" where \"users\".\"id\" = ? intersect select \"users\".\"id\", users.name
from \"users\" left join \"posts\" on \"users\".\"id\" = \"posts\".\" author_id\" where \"users\".\"id\" = ?
intersect select \"id\", \"name\" from \"users\" where \"users\".\"id\" = ? union select \"id\", \"name\"
from \"users\" where \"users\".\"id\" = ?",
params: [ 1, 2, 3, 4, 5, 2 ]
```

Selecting different columns names, with different types or different number of them will result in a type error that if ignored, will result in a runtime error as it will be invalid SQL syntax.

AndriiSherman commented on drizzle-orm/type-tests/pg/set-operators.ts:
```diff
@@ -0,0 +1,194 @@
+
+Expect<Equal<{ userId: number; cityId: number | null }[], typeof exceptAll2Test>>;
+
+// @ts-expect-error - The select on both sites must be the same shape
```
Maybe let's change an error to this. I took it from the Postgres docs, as it seems to be more specific about what was wrong

> All queries in combining statements should return the same number of columns and the corresponding columns should have compatible data type

AndriiSherman commented on drizzle-orm/type-tests/mysql/set-operators.ts:
```diff
@@ -0,0 +1,209 @@
+	db.select().from(cities),
+);
+
+union(
```
Maybe we should have one more test with columns in a different order. In this case, types for columns in the same position will be different, which should be shown as an error. If you haven't implemented it in the types, it would be ideal to add it as well. If it's something that is not possible or has unnecessary complexity, let's just handle it at runtime

```ts
union(
db.select({ id: cities.id, name: cities.name }).from(cities),
db.select({ name: cities.name, id: cities.id }).from(cities),
)
```
Also let's add this case to every possible union/intersect/except implementation

Angelelz replied:
Andrew, unfortunately, implementing this at the type level might not be possible, it would require typescript to have some kind of order on an union type (keyof object). See [this typescript issue](https://github.com/microsoft/TypeScript/issues/13298) for context. I will add a runtime check and throw and error if different key orders are detected.

Angelelz committed:
[All] Completed implementation with requested changes

 - Added coments for type tests

 - Added runtime check that throws an error if the select have keys in different order

 - Added tests for the new error throwing behavior

Angelelz committed:
[All] code cleanup and eslint rules fix

Angelelz committed:
[All] Added offset

Angelelz committed:
[All] added `.as()` method to create subqueries from set operations. Added tests

Angelelz committed:
[All] Moved the query construction logic to the each dialect

dankochetov commented on drizzle-orm/src/mysql-core/dialect.ts:
```diff
@@ -334,6 +335,54 @@ export class MySqlDialect {
+			// which is invalid MySql syntax, Table from one of the SELECTs cannot be used in global ORDER clause
+			for (const orderByUnit of orderBy) {
+				if (is(orderByUnit, MySqlColumn)) {
+					orderByValues.push(sql.raw(orderByUnit.name));
```
You probably need this value escaped in the query?
(also check that all other places where you insert the table/column name directly use `sql.identifier`)
```suggestion
					orderByValues.push(sql.identifier(orderByUnit.name));
```

dankochetov commented on drizzle-orm/src/mysql-core/query-builders/set-operators.ts:
```diff
@@ -0,0 +1,351 @@
+		};
+	}
+
+	private setOperator(
```
For consistency with select builder's `createJoin`
```suggestion
	private createSetOperator(
```

dankochetov commented on drizzle-orm/src/mysql-core/query-builders/set-operators.ts:
```diff
@@ -0,0 +1,351 @@
+	QueryPromise<SelectResult<TSelection, TSelectMode, TNullabilityMap>[]>
+{}
+
+export abstract class MySqlSetOperatorBuilder<
```
Why is this an abstract class? Are there multiple subclasses extending it?

Angelelz replied:
This abstract class is extended by `MySqlSelectQueryBuilderBase` and `MySqlSetOperator` concrete class. I decided this was a better approach than only a concrete class because there were subtle difference in some of the shared methods. Also, I'm avoiding a lot of changes for the `with recursive` implementation. This approach is working great, what do you think would be the benefit of a straight concrete class?

Angelelz replied:
Also, I just realized I made a mistake, and it's probably the reason you're wondering about why is this an abstract class. the `applyMixins` function is supposed to be applied to `MySqlSetOperator` and not the abstract `MySqlSetOperatorBuilder`

dankochetov commented on drizzle-orm/src/mysql-core/query-builders/set-operators.ts:
```diff
@@ -0,0 +1,351 @@
+
+applyMixins(MySqlSetOperatorBuilder, [QueryPromise]);
+
+function setOperator(type: SetOperator, isAll: boolean): <
```
```suggestion
function createSetOperator(type: SetOperator, isAll: boolean): <
```

dankochetov commented on drizzle-orm/src/mysql-core/dialect.ts:
```diff
@@ -334,6 +335,54 @@ export class MySqlDialect {
+				if (is(orderByUnit, MySqlColumn)) {
+					orderByValues.push(sql.raw(orderByUnit.name));
+				} else if (is(orderByUnit, SQL)) {
+					for (let i = 0; i < orderByUnit.queryChunks.length; i++) {
```
Shouldn't this be recursive? SQL's chunks are often SQL instances themselves.

Angelelz replied:
In all the manual and automatic testing I did, this was enough, I couldn't come up with a case where making this recursive was necessary.

dankochetov reviewed (changes requested):
Please change the PR's destination to `beta` instead of `main`. There were some query builder type changes, so you will probably have some merge conflicts.

Also, you'd need to update the `SetOperator` classes according to the changes in other query builders.

Angelelz committed:
Merge branch 'beta' into feat-union

Angelelz committed:
[All] Resolved merge conflicts. Need to fix types

enfipy commented:
Just want to say that I absolutely love this feature and can't wait to see it released! 🔥

Angelelz committed:
[SQLite] fixed types, addressed comments and added new type tests

Angelelz committed:
[SQLite] deleted commented code and improved error readability

Angelelz committed:
[MySql] fixed types, addressed comments and added new type tests

Angelelz committed:
[Pg] fixed types, addressed comments and added new type tests, fixed eslint errors

Angelelz committed:
Merge branch 'beta' into feat-union

Angelelz committed:
[All] fix typescript errors

Angelelz commented:
The implementation is complete.
I included similar types as the rest of the query builders to avoid calling the same method twice.

datner commented:
the hype is real, lets gooooooo

Angelelz committed:
[MySql] Simplification by removing intermediary clases and making the queries mutable

Angelelz committed:
[Pg] Simplification by removing intermediary clases and making the queries mutable

Angelelz committed:
[SQLite] Simplification by removing intermediary clases and making the queries mutable

dankochetov commented on drizzle-orm/src/mysql-core/query-builders/select.types.ts:
```diff
@@ -184,26 +197,51 @@ export interface MySqlSelectHKT extends MySqlSelectHKTBase {
 	>;
 }
 
+export type MySqlSetOperatorExcludedMethods = Exclude<
```
I think it's better to have this as a simple "exclude" list, rather than "double exclude", i.e. to simply list the methods that should be excluded from the set operator result. If a new method is added to the query builder, it's better to have a potential case where a method that should be excluded is included by mistake, rather than a case where a method that should be present is excluded by mistake.

dankochetov replied:
2nd thing is, from what I saw in the docs, only Postgres allows using `limit`, `offset` etc. clauses on the result of the set operator query, i.e. `(select ...) union (select ...) limit X offset Y`. MySQL and SQLite only allow combining set operators with other set operators. So the only methods allowed after a set operator should be set operators in MySQL and SQLite. That should also simplify the SQL generation for these two dialects, since we'd only need to "join" the select queries with a selected set operator kind.

Angelelz replied:
Understood, I thought it would be easier maintaining it the the included instead of the excluded.
In MySql this is a totally valid query:
```sql
(select * from users order by id limit 3 offset 1) union (select * from users where id > 5 order by id limit 2 offset 1) limit 4 offset 1;
```
I will add the excluded methods for SQLite.

Angelelz replied:
Actually, order by, limit and offset are allowed in the "Compound Select Statements" (as they call it in SQLite) as long as it's in the end. But I don't think we have a way to enforce not allowing order by and limit on the selects via types. [See SQLite docs](https://www.sqlite.org/lang_select.html#compound_select_statements)

dankochetov replied:
I see, you are correct. Didn't find any info in MySQL docs about this syntax, but it indeed works. SQLite part is also correct.

Angelelz replied:
For the case of SQLite, what we can do is exclude all the set operators if a limit or order by is used in the left select. But I don't think we'll be able to enforce at the type level that the right select shouldn't have a limit or order by. Would you like me to throw a runtime error if limit or order by is detected in the right select?

dankochetov commented on drizzle-orm/src/mysql-core/query-builders/select.types.ts:
```diff
@@ -260,6 +299,61 @@ export type MySqlSelectQueryBuilder<
 
+export type AnyMySqlSetOperatorInterface = MySqlSetOperatorInterface<any, any, any, any, any, any, any, any, any>;
+
+export interface MySqlSetOperatorInterface<
```
What is the purpose of this interface? What's the functional difference between it and `MySqlSelectBase`?

Angelelz replied:
I needed an easy way to get rid of all the private/protected methods. They were giving me a lot of problems when using the HKTs. With this helper I have access to all the types via `_`.

dankochetov replied:
I see, in that case you might simplify it with something like this I think
```ts
export type MySqlSetOperatorInterface<T extends AnyMySqlSelect> = {
	_: T['_'];
};

export type MySqlSetOperatorWithResult<TResult extends any[]> = {
	_: Omit<AnyMySqlSelect['_'], 'result'> & {
		result: TResult;
	};
};
```

Angelelz replied:
This would indeed work, but the problem is that later on, I need to use the `ValidateShape` type helper. The only way I found I could have the types behave the way I wanted, with the descriptive errors was by [narrowing it first](https://github.com/drizzle-team/drizzle-orm/pull/1218/files#diff-90b01058bcd749df48b13c9b98a75a05d6df28c40cf4b486cdc5ae99e075954bR360-R366). That's why I need this interface like this.

dankochetov commented on drizzle-orm/src/mysql-core/query-builders/select.types.ts:
```diff
@@ -260,6 +299,61 @@ export type MySqlSelectQueryBuilder<
+	TResult extends any[] = SelectResult<TSelection, TSelectMode, TNullabilityMap>[],
+	TSelectedFields extends ColumnsSelection = BuildSubquerySelection<TSelection, TNullabilityMap>,
+> extends
+	Pick<
```
What does `Pick<..., never>` do? Isn't it just excluding everything?

Angelelz replied:
Yes, I was able to get rid of all private/protected methods and just access the types via `_` to enforce the result type.

Angelelz replied:
I just forgot to delete it afterwards

dankochetov commented on drizzle-orm/src/mysql-core/query-builders/select.types.ts:
```diff
@@ -268,3 +362,98 @@ export type MySqlSelect<
+		>
+	: never
+	: TValue;
+
```
Use the `_` property reference instead of inference whenever possible, it's cleaner and was specifically designed for such cases.
```suggestion
> = ValidateShape<
	TValue['_']['result'][number],
	TResult[number],
	TypedQueryBuilder<any, TValue['_']['result']>
>;
```

Angelelz replied:
I wish I had thought of that. You have no idea how long I spent with this stupid type. I'll commit in my code to run all the tests again

Angelelz replied:
For some reason if I don't narrow the type the way I implemented it, the ValidateShape won't pick it up and won't show the error. I added your suggested change, but the type errors don't show. If you can set it up and check the type-tests let me know. I did however simplified it a little based on your suggestion.

Angelelz replied:
I think I know the reason. `TValue` extends `MySqlSetOperatorWithResult<TResult>` and the `TResult` is pulled from the type we actually want to validate against. When we compare `TValue['_']['result'][number]` against `TResult[number]` the compiler is basically comparing the same type against itself.
That's why narrowing _**and inferring**_ is necessary for the ValidateShape to work properly.

dankochetov replied:
Makes sense, I might look into the types later to see if they can be further optimized. Type performance is our bottleneck in general, so we have to be careful not to introduce too much type checking overhead.

dankochetov commented on drizzle-orm/src/utils.ts:
```diff
@@ -194,6 +211,12 @@ export interface DrizzleConfig<TSchema extends Record<string, unknown> = Record<
 	logger?: boolean | Logger;
 	schema?: TSchema;
 }
+export type ValidateShape<T, ValidShape, TResult = T> = T extends ValidShape
```
Set operators allow for the column names to be different between the queries, as long as the data types are the same. I.e. this query, for example, is correct:
```sql
select 1 as a union select 2 as b;
```
Therefore, it should be possible to have a query like this:
```ts
db
  .select({
    a: sql`1`
  })
  .union(
    db.select({
      b: sql`2`
    }),
  )
```

I don't think it's possible to validate this on type level, therefore I don't think we can make use of the `ValidateShape` type. Corresponding changes should be also made to the runtime checks. We should also have a test that covers such cases.

Angelelz replied:
This will be by far the hardest change as the whole type safety I built was around the idea of maintaining the same column names.
I'd argue we should keep it the way it is currently implemented. I think allowing different column names will introduce confusion because, the column names are only used in the JS/TS side during mapping.
For example, in the current implementation the following is allowed:
```ts
db
  .select({
    a: sql`1`.as("a")
  })
  .union(
    db.select({
      a: sql`2`.as("b")
    }),
  )
```
Which will produce the following sql:
```sql
select 1 as a union select 2 as b;
```
and is valid sql.

Angelelz replied:
If we allow different `SelectedFields` keys, what should the user expect the end type to be?
Furthermore, drizzle has historically not cared about the underline *SQL* column names and maps the results to the `SelectedFields` keys, whatever they are. For example:
```ts
await db.select({ someKey: users.name }).from(users)
```
Has always been of type:
```ts
{ someKey: string }[]
```
And not
```ts
{ name: string }[]
```

Angelelz committed:
[MySql] Types simplification from review and fixed tests for western hemisphere

Angelelz committed:
[Pg] Types simplification from review

Angelelz committed:
[SQLite] Types simplification from review

Angelelz committed:
[MySql] fixed failing tests based on query order

Angelelz commented on drizzle-orm/src/mysql-core/query-builders/select.ts:
```diff
@@ -260,6 +268,61 @@ export abstract class MySqlSelectQueryBuilderBase<
+				>;
+
+			if (!haveSameKeys(this.getSelectedFields(), rightSelect.getSelectedFields())) {
+				throw new Error(
```
This is the code that throws if different `SelectedField` keys are detected. We might want to think if we want to keep this code. @dankochetov

dankochetov replied:
Let's leave it like this for now, it's better to gather feedback first.

dankochetov reviewed (approved):
LET'S GO THE HYPE IS REAL
